lib/Parse/ParsePattern.cpp

```diff
@@ -821,52 +821,102 @@ Parser::parseFunctionSignature(Identifier SimpleName,
-    if (Tok.isAny(tok::kw_throw, tok::kw_try)) {
-      diagnose(Tok, diag::throw_in_function_type)
-        .fixItReplace(Tok.getLoc(), "throws");
-    }
+ParserStatus Parser::parseEffectsSpecifiers(SourceLoc existingArrowLoc,
+                                            SourceLoc &asyncLoc,
+                                            SourceLoc &throwsLoc,
+                                            bool *rethrows) {
+  ParserStatus status;
 
-    StringRef keyword = Tok.getText();
-    throwsLoc = consumeToken();
+  while (true) {
+    // 'async'
+    if (shouldParseExperimentalConcurrency() &&
+        Tok.isContextualKeyword("async")) {
 
-    if (existingArrowLoc.isValid()) {
-      diagnose(throwsLoc, diag::async_or_throws_in_wrong_position,
-               rethrows ? (*rethrows ? 1 : 0) : 0)
-        .fixItRemove(throwsLoc)
-        .fixItInsert(existingArrowLoc, (keyword + " ").str());
+      if (asyncLoc.isValid()) {
+        diagnose(Tok, diag::duplicate_effect_specifier, Tok.getText())
+            .fixItRemove(Tok.getLoc());
+      } else if (existingArrowLoc.isValid()) {
+        SourceLoc insertLoc = existingArrowLoc;
+        if (throwsLoc.isValid() &&
+            SourceMgr.isBeforeInBuffer(throwsLoc, insertLoc))
+          insertLoc = throwsLoc;
+        diagnose(Tok, diag::async_or_throws_in_wrong_position, 2)
+            .fixItRemove(Tok.getLoc())
+            .fixItInsert(insertLoc, "async ");
+      } else if (throwsLoc.isValid()) {
+        // 'async' cannot be after 'throws'.
+        assert(existingArrowLoc.isInvalid());
+        diagnose(Tok, diag::async_after_throws, rethrows && *rethrows)
+            .fixItRemove(Tok.getLoc())
+            .fixItInsert(throwsLoc, "async ");
+      }
+      if (asyncLoc.isInvalid())
+        asyncLoc = Tok.getLoc();
+      consumeToken();
+      continue;
     }
 
-    if (shouldParseExperimentalConcurrency() &&
-        Tok.isContextualKeyword("async")) {
-      asyncLoc = consumeToken();
+    // 'throws'/'rethrows', or diagnose 'throw'/'try'.
+    if (Tok.isAny(tok::kw_throws, tok::kw_rethrows) ||
+        (Tok.isAny(tok::kw_throw, tok::kw_try) && !Tok.isAtStartOfLine())) {
+      bool isRethrows = Tok.is(tok::kw_rethrows);
+
+      if (throwsLoc.isValid()) {
+        diagnose(Tok, diag::duplicate_effect_specifier, Tok.getText())
+            .fixItRemove(Tok.getLoc());
+      } else if (Tok.isAny(tok::kw_throw, tok::kw_try)) {
+        // Replace 'throw' or 'try' with 'throws'.
+        diagnose(Tok, diag::throw_in_function_type)
+            .fixItReplace(Tok.getLoc(), "throws");
+      } else if (!rethrows && isRethrows) {
+        // Replace 'rethrows' with 'throws' unless it's allowed.
+        diagnose(Tok, diag::rethrowing_function_type)
+            .fixItReplace(Tok.getLoc(), "throws");
+      } else if (existingArrowLoc.isValid()) {
+        diagnose(Tok, diag::async_or_throws_in_wrong_position,
+                 rethrows ? (isRethrows ? 1 : 0) : 0)
```

Comment:
Is there a reason to check `rethrows` here? Seems better for the diagnostic to match the token if even rethrows is not actually allowed, since the fixit is going to keep the `rethrows` anyway.

Reply:
I just kept the existing code, but I agree. Changed to use the token text as is.